File: package/main/src/Async/debounceAsync.ts
export interface DebouncedAsyncFunction<A extends unknown[], R> {
  (...arguments_: A): Promise<R>;
  cancel: () => void;
}

/**
 * Creates a debounced async function. Subsequent calls within `wait` ms reset
 * the timer and share a single resolution; the latest arguments win.
 *
 * @template A - Argument tuple type
 * @template R - Resolved value type
 * @param {(...args: A) => Promise<R>} function_ - Async function to debounce
 * @param {number} wait - Debounce window in milliseconds
 * @returns {DebouncedAsyncFunction<A, R>} Debounced wrapper with cancel support
 * @example
 * const search = debounceAsync(query, 300);
 * await search("foo");
 */
export const debounceAsync = <A extends unknown[], R>(
  function_: (...arguments_: A) => Promise<R>,
  wait: number,
): DebouncedAsyncFunction<A, R> => {
  let timer: ReturnType<typeof setTimeout> | undefined;
  let pendingResolvers: ((value: R) => void)[] = [];
  let pendingRejecters: ((reason: unknown) => void)[] = [];

  const flushPromises = (): {
    resolvers: ((value: R) => void)[];
    rejecters: ((reason: unknown) => void)[];
  } => {
    const resolvers = pendingResolvers;
    const rejecters = pendingRejecters;
    pendingResolvers = [];
    pendingRejecters = [];
    return { resolvers, rejecters };
  };

  const debounced = ((...arguments_: A): Promise<R> => {
    if (timer) {
      clearTimeout(timer);
    }
    return new Promise<R>((resolve, reject) => {
      pendingResolvers.push(resolve);
      pendingRejecters.push(reject);
      timer = setTimeout(() => {
        timer = undefined;
        const { resolvers, rejecters } = flushPromises();
        function_(...arguments_).then(
          (value) => {
            for (const resolver of resolvers) {
              resolver(value);
            }
          },
          (error) => {
            for (const rejecter of rejecters) {
              rejecter(error);
            }
          },
        );
      }, wait);
    });
  }) as DebouncedAsyncFunction<A, R>;

  debounced.cancel = (): void => {
    if (timer) {
      clearTimeout(timer);
      timer = undefined;
    }
    const { rejecters } = flushPromises();
    for (const rejecter of rejecters) {
      rejecter(new Error("debounceAsync cancelled"));
    }
  };

  return debounced;
};

File: package/umt_rust/src/async_util/debounce_async.rs
//! A debounced wrapper around an operation.
//!
//! Mirrors the TypeScript `debounceAsync`: calls made within the `wait` window
//! reset the timer and share a single underlying invocation, the latest
//! arguments win, and every caller receives the same result. `cancel` rejects
//! every pending caller with `debounceAsync cancelled`.
//!
//! Each caller receives a [`Receiver`] that delivers the settled
//! `Result<R, String>` once the debounced invocation completes (or is
//! cancelled). Block on it with `recv` to obtain the value, matching the
//! `await` of the TypeScript version.

use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::mpsc::{self, Receiver, Sender};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

type Pending<R> = Vec<Sender<Result<R, String>>>;

struct DebounceState<A, R> {
    generation: AtomicU64,
    wait: u64,
    pending: Mutex<Pending<R>>,
    latest_args: Mutex<Option<A>>,
}

/// A debounced wrapper produced by [`umt_debounce_async`].
pub struct DebouncedAsync<A, R, F> {
    state: Arc<DebounceState<A, R>>,
    function: Arc<F>,
}

impl<A, R, F> DebouncedAsync<A, R, F>
where
    A: Clone + Send + 'static,
    R: Clone + Send + 'static,
    F: Fn(A) -> Result<R, String> + Send + Sync + 'static,
{
    /// Schedules an invocation with the given arguments, resetting the debounce
    /// timer. Returns a receiver that yields the shared result once the window
    /// elapses, or the cancellation error if `cancel` is called first.
    pub fn call(&self, args: A) -> Receiver<Result<R, String>> {
        let (sender, receiver) = mpsc::channel();
        {
            let mut pending = self.state.pending.lock().unwrap();
            pending.push(sender);
        }
        *self.state.latest_args.lock().unwrap() = Some(args);

        let generation = self.state.generation.fetch_add(1, Ordering::SeqCst) + 1;
        let state = Arc::clone(&self.state);
        let function = Arc::clone(&self.function);
        thread::spawn(move || {
            thread::sleep(Duration::from_millis(state.wait));
            if state.generation.load(Ordering::SeqCst) != generation {
                return;
            }
            let resolvers: Pending<R> = std::mem::take(&mut *state.pending.lock().unwrap());
            let args = state.latest_args.lock().unwrap().clone();
            let Some(args) = args else {
                return;
            };
            let outcome = function(args);
            for resolver in resolvers {
                let _ = resolver.send(outcome.clone());
            }
        });

        receiver
    }

    /// Cancels any pending invocation, rejecting every waiting caller with
    /// `debounceAsync cancelled`. A no-op when nothing is pending.
    pub fn cancel(&self) {
        self.state.generation.fetch_add(1, Ordering::SeqCst);
        let rejecters: Pending<R> = std::mem::take(&mut *self.state.pending.lock().unwrap());
        for rejecter in rejecters {
            let _ = rejecter.send(Err("debounceAsync cancelled".to_string()));
        }
    }
}

/// Creates a debounced wrapper around `function_`.
///
/// # Arguments
///
/// * `function_` - The operation to debounce; receives the latest arguments and
///   returns `Result<R, String>`
/// * `wait` - The debounce window in milliseconds
///
/// # Returns
///
/// A [`DebouncedAsync`] exposing `call` and `cancel`.
///
/// # Examples
///
/// ```
/// use umt_rust::async_util::umt_debounce_async;
///
/// let debounced = umt_debounce_async(|value: i32| Ok::<i32, String>(value * 2), 5);
/// let a = debounced.call(1);
/// let b = debounced.call(2);
/// let c = debounced.call(3);
/// assert_eq!(a.recv().unwrap(), Ok(6));
/// assert_eq!(b.recv().unwrap(), Ok(6));
/// assert_eq!(c.recv().unwrap(), Ok(6));
/// ```
pub fn umt_debounce_async<A, R, F>(function_: F, wait: u64) -> DebouncedAsync<A, R, F>
where
    A: Clone + Send + 'static,
    R: Clone + Send + 'static,
    F: Fn(A) -> Result<R, String> + Send + Sync + 'static,
{
    DebouncedAsync {
        state: Arc::new(DebounceState {
            generation: AtomicU64::new(0),
            wait,
            pending: Mutex::new(Vec::new()),
            latest_args: Mutex::new(None),
        }),
        function: Arc::new(function_),
    }
}
